lib/URL-impl.js

```diff
@@ -0,0 +1,214 @@
+"use strict";
+const usm = require("./url-state-machine");
+
+exports.implementation = class URLImpl {
+  constructor(constructorArgs) {
+    const url = constructorArgs[0];
+    const base = constructorArgs[1];
+
+    let parsedBase = null;
+    if (base !== undefined) {
+      parsedBase = usm.basicURLParse(base);
+      if (parsedBase.failure) {
+        throw new TypeError("Invalid base URL");
+      }
+    }
+
+    const parsedURL = usm.basicURLParse(url, { baseURL: parsedBase });
+    if (parsedURL.failure) {
+      throw new TypeError("Invalid URL");
+    }
+
+    this._url = parsedURL;
+
+    // TODO: query stuff
+  }
+
+  static domainToASCII(domain) {
+    try {
```

Comment:
Looks like parseHost can't throw anymore so I don't think we need this anymore, right?